File: app/src/main/java/com/example/expensetrcaker/viewmodel/TransactionViewModel.java
package com.example.expensetrcaker.viewmodel;

import android.app.Application;
import android.graphics.Color;
import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import com.example.expensetrcaker.ai.SpendingAI;
import com.example.expensetrcaker.ai.SpendingPersonality;
import com.example.expensetrcaker.data.AppDatabase;
import com.example.expensetrcaker.data.BudgetDao;
import com.example.expensetrcaker.data.BudgetEntity;
import com.example.expensetrcaker.data.TransactionDao;
import com.example.expensetrcaker.data.TransactionEntity;
import com.example.expensetrcaker.data.WalletDao;
import com.example.expensetrcaker.data.WalletEntity;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class TransactionViewModel extends AndroidViewModel {
    private final TransactionDao transactionDao;
    private final BudgetDao budgetDao;
    private final WalletDao walletDao;
    private final ExecutorService executorService;

    private final LiveData<List<TransactionEntity>> allTransactions;
    private final LiveData<List<BudgetEntity>> allBudgets;
    private final LiveData<List<WalletEntity>> allWallets;
    private final LiveData<Double> totalIncome;
    private final LiveData<Double> totalExpense;

    public TransactionViewModel(@NonNull Application application) {
        super(application);
        AppDatabase db = AppDatabase.getDatabase(application);
        transactionDao = db.transactionDao();
        budgetDao = db.budgetDao();
        walletDao = db.walletDao();
        executorService = Executors.newSingleThreadExecutor();

        allTransactions = transactionDao.getAllTransactions();
        allBudgets = budgetDao.getAllBudgets();
        allWallets = walletDao.getAllWallets();
        totalIncome = transactionDao.getTotalIncome();
        totalExpense = transactionDao.getTotalExpense();

        // Initialize default wallet if empty
        allWallets.observeForever(wallets -> {
            if (wallets != null && wallets.isEmpty()) {
                executorService.execute(() -> {
                    walletDao.insertWallet(new WalletEntity(
                            "Main Wallet",
                            5000.0,
                            "Mastercard",
                            Color.parseColor("#8E2DE2"),
                            Color.parseColor("#4A00E0")));
                });
            }
        });
    }

    public LiveData<List<TransactionEntity>> getAllTransactions() {
        return allTransactions;
    }

    public LiveData<List<BudgetEntity>> getAllBudgets() {
        return allBudgets;
    }

    public LiveData<List<WalletEntity>> getAllWallets() {
        return allWallets;
    }

    public LiveData<Double> getTotalIncome() {
        return Transformations.map(totalIncome, val -> val != null ? val : 0.0);
    }

    public LiveData<Double> getTotalExpense() {
        return Transformations.map(totalExpense, val -> val != null ? val : 0.0);
    }

    public LiveData<Double> getTotalBalance() {
        return Transformations.map(allWallets, wallets -> {
            double sum = 0;
            if (wallets != null) {
                for (WalletEntity w : wallets)
                    sum += w.getBalance();
            }
            return sum;
        });
    }

    public LiveData<SpendingPersonality> getPersonality() {
        return Transformations.map(allTransactions, transactions -> SpendingAI.calculatePersonality(transactions));
    }

    public void addTransaction(String title, double amount, String category, boolean isExpense, long walletId) {
        executorService.execute(() -> {
            transactionDao.insertTransaction(new TransactionEntity(
                    title, amount, category, System.currentTimeMillis(), isExpense, walletId));
            double balanceChange = isExpense ? -amount : amount;
            walletDao.updateBalance(walletId, balanceChange);
        });
    }

    public void deleteTransaction(TransactionEntity transaction) {
        executorService.execute(() -> {
            transactionDao.deleteTransaction(transaction);
            double balanceChange = transaction.isExpense() ? transaction.getAmount() : -transaction.getAmount();
            walletDao.updateBalance(transaction.getWalletId(), balanceChange);
        });
    }

    public void upsertBudget(String category, double limit) {
        executorService.execute(() -> {
            budgetDao.upsertBudget(new BudgetEntity(category, limit, "Monthly"));
        });
    }

    public void addWallet(String name, double balance, String bank, String colorStart, String colorEnd) {
        executorService.execute(() -> {
            walletDao.insertWallet(new WalletEntity(
                    name, balance, bank, Color.parseColor(colorStart), Color.parseColor(colorEnd)));
        });
    }

    public void deleteWallet(WalletEntity wallet) {
        executorService.execute(() -> {
            walletDao.deleteWallet(wallet);
        });
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        executorService.shutdown();
    }
}

File: app/src/main/java/com/example/expensetrcaker/viewmodel/TransactionViewModel.kt
package com.example.expensetrcaker.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.example.expensetrcaker.data.AppDatabase
import com.example.expensetrcaker.data.BudgetEntity
import com.example.expensetrcaker.data.TransactionEntity
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch

class TransactionViewModel(application: Application) : AndroidViewModel(application) {
    private val dao = AppDatabase.getDatabase(application).transactionDao()

    val allTransactions: StateFlow<List<TransactionEntity>> = dao.getAllTransactions()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    val totalIncome: StateFlow<Double> = dao.getTotalIncome()
        .map { it ?: 0.0 }
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), 0.0)

    val totalExpense: StateFlow<Double> = dao.getTotalExpense()
        .map { it ?: 0.0 }
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), 0.0)

    val balance: StateFlow<Double> = totalIncome.map { it } // Simplified for now
        // Usually balance = income - expense
        // But we'll handle the math in the UI or here.
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), 0.0)

    private val budgetDao = AppDatabase.getDatabase(application).budgetDao()
    val allBudgets = budgetDao.getAllBudgets().stateIn(viewModelScope, SharingStarted.Lazily, emptyList())

    private val walletDao = AppDatabase.getDatabase(application).walletDao()
    val allWallets = walletDao.getAllWallets().stateIn(viewModelScope, SharingStarted.Lazily, emptyList())

    val totalBalance = allWallets.map { wallets ->
        wallets.sumOf { it.balance }
    }.stateIn(viewModelScope, SharingStarted.Lazily, 0.0)

    val personality = allTransactions.map { transactions ->
        com.example.expensetrcaker.ai.SpendingAI.calculatePersonality(transactions)
    }.stateIn(viewModelScope, SharingStarted.Lazily, com.example.expensetrcaker.ai.SpendingAI.calculatePersonality(emptyList()))

    init {
        // Initialize common wallets if empty
        viewModelScope.launch {
            walletDao.getAllWallets().collect {
                if (it.isEmpty()) {
                    walletDao.insertWallet(com.example.expensetrcaker.data.WalletEntity(
                        name = "Main Wallet",
                        balance = 5000.0,
                        bank = "Mastercard",
                        colorStart = android.graphics.Color.parseColor("#8E2DE2"),
                        colorEnd = android.graphics.Color.parseColor("#4A00E0")
                    ))
                }
            }
        }
    }

    fun addTransaction(title: String, amount: Double, category: String, isExpense: Boolean, walletId: Long = 1) {
        viewModelScope.launch {
            dao.insertTransaction(
                com.example.expensetrcaker.data.TransactionEntity(
                    title = title,
                    amount = amount,
                    date = System.currentTimeMillis(),
                    category = category,
                    isExpense = isExpense,
                    walletId = walletId
                )
            )
            // Update Wallet Balance
            val balanceChange = if (isExpense) -amount else amount
            walletDao.updateBalance(walletId, balanceChange)
        }
    }

    fun deleteTransaction(transaction: com.example.expensetrcaker.data.TransactionEntity) {
        viewModelScope.launch {
            dao.deleteTransaction(transaction)
            // Reverse Wallet Balance
            val balanceChange = if (transaction.isExpense) transaction.amount else -transaction.amount
            walletDao.updateBalance(transaction.walletId, balanceChange)
        }
    }

    fun upsertBudget(category: String, limit: Double) {
        viewModelScope.launch {
            budgetDao.upsertBudget(com.example.expensetrcaker.data.BudgetEntity(category, limit))
        }
    }

    fun addWallet(name: String, balance: Double, bank: String, colorStart: String, colorEnd: String) {
        viewModelScope.launch {
            walletDao.insertWallet(com.example.expensetrcaker.data.WalletEntity(
                name = name,
                balance = balance,
                bank = bank,
                colorStart = android.graphics.Color.parseColor(colorStart),
                colorEnd = android.graphics.Color.parseColor(colorEnd)
            ))
        }
    }

    fun deleteWallet(wallet: com.example.expensetrcaker.data.WalletEntity) {
        viewModelScope.launch {
            walletDao.deleteWallet(wallet)
        }
    }
}
